File: Data Structures & Algorithms/level-order-traversal-of-binary-tree/submission-2.ts
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number[][]}
     */
    levelOrder(root: TreeNode | null): number[][] {
        if (!root) return [];
        const res = [];
        const queue = [root];
        let front = 0;
        while (front < queue.length) {
            let level = [];
            let levelSize = queue.length - front;
            for (let i = 0; i < levelSize; i++) {
                //pushing current value in level
                let node = queue[front++];
                level.push(node.val);

                if (node.left) {
                    queue.push(node.left);
                }
                if (node.right) {
                    queue.push(node.right);
                }
            }
            res.push(level);
        }
        return res;
    }
}

File: Data Structures & Algorithms/binary-tree-from-preorder-and-inorder-traversal/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if (not preorder) or (not inorder):
            return None

        inoderMap = {value: i for i, value in enumerate(inorder)}
        self.preorderIndex = 0
        def dfs(left, right):
            if left > right:
                return None
            rootVal = preorder[self.preorderIndex]
            self.preorderIndex += 1
            root = TreeNode(rootVal)
            mid = inoderMap[rootVal]
            root.left = dfs(left, mid - 1)
            root.right = dfs(mid+1, right)
            return root

        return dfs(0, len(inorder)-1)
        